lib/chef/provider/service/systemd.rb

```diff
@@ -76,6 +76,30 @@ def define_resource_requirements
     end
   end
 
+  def systemd_service_status
+    @service_status ||= begin
```

Comment:
I worry a bit about the memoization here.

There should be one provider class instantiation per resource, I think, so in theory I don't think this will break:
```
service 'foo' do  # we check, see it's running
  action :stop
end
# something
service 'foo' do   # we need to check again, since it's now not running
  action :start
end
```
So I think that case is OK, but we probably want a test for it.

What about if we call this during `define_resource_requirements`? The wiring up of dynamic providers is a bit fuzzy to me, these days, but @lamont-granquist does that run during compile time? If so, we'll definitely break the entire world, that'd be like depending on Ohai to figure out if your service is running.

We also want a test around that case.

Reply:
I didn't test cloned resources, but I did test multiple actions on a single resource, with the very funny:
```
service 'chronyd' do
  action [:start, :stop, :start, :stop, :start, :start, :start]
end
```

This still works; i.e. each `action` re-evaluates the state of the service, and idempotently fixes it when necessary.

Reply:
I assume you mean tested manually yourself. But that's not why I want a test. I want a test so that in the future when someone innocently adds a call to this function from the wrong place and or we change up some wiring, we don't then break this code as a side effect.

You can probably add a similar test to the above in the `integration` test directory where you can just write a simple recipe. But one way or another, if we're gonna memoize stuff, we shouldn't merge this without a test to ensure we never have a stale memo.

Reply:
the ivar and the memoization helper should be really named the same thing, so either service_status or systemd_service_status.

Reply:
so the memoization approach should be fine, every time an action is called on a resource it gets a new provider ("action class") with fresh state, so repeated resources or array actions will be fine.  even `run_action(:whatever)` on a saved resource will flush the provider.  we do a lot of memoization in other resources.

Reply:
@jaymzh - since every `action` gets a new provider (per Lamont and corroborated by my testing), I don't think we need a test to cover this case.  Browsing through the `integration` dir, we could mock out a resource running multiple actions, but since each provider is a separate instance, that wouldn't be testing the memoization in any way.